src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
-//! The type checker
-
 use crate::syntax::*;
+use thiserror::Error;
+
+/// Errors that can occur during type checking
+#[derive(Error, Debug)]
+pub enum TypeError {
+    /// Type mismatch between expected and actual types
+    #[error("Type mismatch: expected {expected:?}, found {actual:?}")]
+    TypeMismatch {
+        /// The type that was expected
+        expected: Type,
+        /// The type that was actually found
+        actual: Type,
+    },
+    /// Variable not found in the environment
+    #[error("Undefined variable: '{0}'")]
+    UndefinedVariable(Id),
+    /// Function not found
+    #[error("Undefined function: '{0}'")]
+    UndefinedFunction(Id),
+    /// Symbol not in alphabet
+    #[error("Symbol '{0}' is not in the alphabet")]
+    SymbolNotInAlphabet(char),
+    /// Wrong number of arguments
+    #[error("Wrong number of arguments: expected {expected}, actual {actual}")]
+    WrongNumberOfArguments {
+        /// Expected number of arguments
+        expected: usize,
+        /// Actual number of arguments
+        actual: usize,
+    },
+    /// Empty match cases
+    #[error("Empty match cases")]
+    EmptyMatchCases,
+    /// Type mismatch between pattern and scrutinee
+    #[error("Type mismatch between pattern and scrutinee")]
+    MismatchPatternScrutinee {
+        /// Expected type
+        expected: Type,
+        /// Actual pattern
+        actual: Pattern,
+    },
+}
+
+/// Type context for type checking
+pub struct TypeCtx<'prog> {
+    /// Type environment
+    env: Map<Id, Type>,
+    /// Function environment
+    funcs: &'prog Map<Id, Function>,
+}
+
+/// Type check an expression in a given type context
+pub fn typeck_expr(expr: &Expr, ctx: &TypeCtx) -> Result<Type, TypeError> {
+    match expr {
+        // Variables are always of the type of the variable in the environment
+        Expr::Var(id) => {
+            // Get the type of the variable from the environment
+            ctx.env
+                .get(id)
+                .cloned()
+                .ok_or(TypeError::UndefinedVariable(*id))
+        }
+        // Boolean literals are always of type BoolT
+        Expr::Bool(_) => Ok(Type::BoolT),
+        // Numeric literals are always of the type they are given
+        Expr::Num(_n, t) => Ok(t.clone()),
+        // Function calls are always of the type of the function
+        Expr::Sym(_) => Ok(Type::SymT),
+        // Binary operations are always of the type of the left hand side
+        Expr::BinOp { lhs, op, rhs } => {
+            let lhs_type = typeck_expr(lhs, ctx)?;
+            let rhs_type = typeck_expr(rhs, ctx)?;
+            match op {
+                // Arithmetic operations are always of type NumT(0..1)
+                BOp::Add | BOp::Sub | BOp::Mul | BOp::Div | BOp::Rem | BOp::Shl | BOp::Shr => {
+                    if let Type::NumT(_) = lhs_type {
```

Comment:
you can combine this and the next if (you can even combine that with the pattern match above). E.g., something like: `lhs_type == rhs_type && matches!(lhs_type, Type::NumT(_))`.